### ISSUE TITLE
CS8602 false positive warning

### ISSUE THREAD

**Issue by taori** [dotnet/roslyn#64762]
**Version Used**: 

	<PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="6.0.0"
	<PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.3"
	<PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.3.1" 

## VS Info
<details>
  <summary>Expand</summary>
  
Microsoft Visual Studio Community 2022
Version 17.3.5
VisualStudio.17.Release/17.3.5+32922.545
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Community

ASP.NET and Web Tools   17.3.376.3011
ASP.NET and Web Tools

Azure App Service Tools v3.0.0   17.3.376.3011
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   17.3.376.3011
Azure Functions and Web Jobs Tools

C# Tools   4.3.0-3.22470.13+80a8ce8d5fdb9ceda4101e2acb8e8eb7be4ebcea
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.2.6 (master@34d6af2)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Markdown Editor v2   2.0.130
A full featured Markdown editor with live preview and syntax highlighting. Supports GitHub flavored Markdown.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Mono Debugging for Visual Studio   17.3.20 (3f4cb00)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   6.3.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

Razor (ASP.NET Core)   17.0.0.2232702+e1d654e792aa2fe6646a6935bcca80ff0aff4387
Provides languages services for ASP.NET Core Razor.

Solution Colors   1.1.44
Allows you to associate a color with a solution and display it in various locations within Visual Studio. Inspired by the Peacock extension for VS Code.

SQL Server Data Tools   17.0.62207.04100
Microsoft SQL Server Data Tools

Syntax Visualizer   1.0
An extension for visualizing Roslyn SyntaxTrees.

TypeScript Tools   17.0.10701.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   4.3.0-3.22470.13+80a8ce8d5fdb9ceda4101e2acb8e8eb7be4ebcea
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   17.1.0-beta.22363.4+1b94f89d4d1f41f20f9be73c76f4b229d4e49078
Microsoft Visual F# Tools

Visual Studio IntelliCode   2.2
AI-assisted development for Visual Studio.

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

VSPackage Extension   1.0
VSPackage Visual Studio Extension Detailed Info

Web Compiler   1.14.8
Compiler for LESS, Sass and CoffeeScript files

Xamarin   17.3.0.308 (d17-3@56b9f7e)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   17.3.0.217 (remotes/origin/d17-3@b81db1dcc0)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   17.2.28 (9b3d244)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   13.0.0.0 (d17-3/030cd63)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: dffa5ab
    Java.Interop: xamarin/java.interop/d17-3@7716ae53
    SQLite: xamarin/sqlite/3.38.5@df4deab
    Xamarin.Android Tools: xamarin/xamarin-android-tools/main@14076a6

Xamarin.iOS and Xamarin.Mac SDK   16.0.0.75 (e4f48bed7)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

</details>

## Steps to Reproduce:

![grafik](https://user-images.githubusercontent.com/5545184/196048484-ca943b0e-398c-4cd0-ab83-702853afb62b.png)

Code:

```cs
private static INamedTypeSymbol? GetClassSymbol(GeneratorAttributeSyntaxContext context, CancellationToken cancellationToken)
	{
		if (context.TargetSymbol is INamedTypeSymbol classSymbol)
		{
			var hierarchyAttribute = context.Attributes
				.Select(static d => GetHierarchyData(d))
				.Where(static d => d is not null);

			foreach (var attributeData in hierarchyAttribute)
			{
				if (attributeData is {Data.ConstructorArguments.Length: > 0})
				{

				}

				if (attributeData.Data.ConstructorArguments.Length > 0)
				{

				}
			}
		}

		return default;
	}

	private static HierarchyData? GetHierarchyData(AttributeData attributeData)
	{
		if (attributeData is {AttributeClass.MetadataName: { } metadataName})
		{
			if (metadataName is Constants.ChildAttributeName)
				return new HierarchyData(attributeData, true, false);
			if (metadataName is Constants.ParentAttributeName)
				return new HierarchyData(attributeData, true, false);
		}

		return default;
	}

	private record HierarchyData(AttributeData Data, bool ChildAttribute, bool ParentAttribute);
```

**Expected Behavior**:
no dereference warning

**Actual Behavior**:
dereference warning 

Additional context:

if i do `.Cast<HierarchyData>` after checking for null it works fine, but ideally the prior nullcheck flows into the type inference for `var hierarchyAttribute` so it is infered as `IEnumerable<HierarchyData>?` instead of `IEnumerable<HierarchyData?>?`

**Comment by CyrusNajmabadi** [comment]
Closing out.  THere is no rule in the language currently that flows information in a lambda into the type system like this.  There are tracking issues in csharplang looking to see if such a thing would be possible.  However, currently, Roslyn (the language impl) is behaving as spec'ed.
